cli/tests/extensions.rs

```diff
@@ -282,6 +284,97 @@ fn conflicting_extension_name_is_filtered() {
     assert!(output.contains("extension was filtered out"));
 }
 
+// Test that the rules of the module loader are obeyed:
+// - Both TypeScript and JavaScript are supported.
+// - Files under $XDG_DATA_HOME/phylum/extensions may be imported.
+// - Remote URLs under https://deno.land are supported -- i.e., the Deno's standard library.
+// - No other URLs are supported.
+//   - We explicitly test that a https:// url which is not under `deno.land` is rejected.
+//   - We explicitly test that a directory traversal attempt is rejected.
+//
+// These tests are based on the fixtures under `fixtures/module-import-extension`.
```

Comment:
Wouldn't it be easier to have individual tests for all these? That would also mean we don't need a massive comment explaining things.

Reply:
This is one thing I've considered anyway, but shouldn't we maybe have some abstractions to just run generic extension code? Having an abstraction to install & run from a path would be nice, but I feel like just having a helper that builds extensions from a string would clean things up significantly.

> Furthermore, I think having the complete set of rules there as a comment makes it clear what are the boundaries that we are testing, and I feel the current solution is a good "logical box".

I think each individual rule would be a good test, combining all of them into one, doesn't make sense to me. This is clearly running multiple integration tests, we're running the binary multiple times and testing their output, so why throw it into one test case?

Reply:
> This is one thing I've considered anyway, but shouldn't we maybe have some abstractions to just run generic extension code? Having an abstraction to install & run from a path would be nice, but I feel like just having a helper that builds extensions from a string would clean things up significantly.

Indeed, there is a lot of repetition. Let's create an issue for it and address it separately, as it is a big enough diff to deserve its own PR.

> I think each individual rule would be a good test, combining all of them into one, doesn't make sense to me. This is clearly running multiple integration tests, we're running the binary multiple times and testing their output, so why throw it into one test case?

I think those assertions are conceptually tightly coupled, don't really stand out on their own even if the form in which they are tested is the execution of a binary, and should be dealt with in bulk -- i.e. if I change one assertion in isolation I want to be aware of the fact that there is a bigger set of behaviors I need to be aware of, and I feel like multiple asserts in the same function is what makes the most sense.

Reply:
That's kind of a slippery slope argument -- it's not like our only options are as extreme as _one assert per test_ vs _one single test for everything_, and the current solution is the middle ground grouping that makes the most sense to me on a logical standpoint.

We're not testing _the module loader_, but _the ruleset of the module loader_ whose rules don't make any sense in isolation: asserting that we accept `deno.land` hosts on its own and then asserting, in a separate logical context, that we don't accept some other remote URLs would be outright confusing to me.

At the same time, filesystem paths and remote URLs are still `Url` objects and we're testing for rules on `Url`s.

If we incur into a bug that's either related to URLs or local files, then we should rather introduce a new, separate test for controlling that individual behavior, _or_ add a stricter formal rule that all `Url`s have to abide to, within this test.

A compromise I'm willing to make would be breaking up the test, and grouping the resulting tests in a dedicated `mod {}` block. The important thing is that the tight interdependence of those assertions is made evident.